src/loc.rs

```diff
@@ -0,0 +1,466 @@
+use endianity::{Endianity, EndianBuf};
+use fallible_iterator::FallibleIterator;
+use parser::{Error, Result, parse_u16, take};
+use ranges::Range;
+use std::marker::PhantomData;
+
+/// An offset into the `.debug_loc` section.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct DebugLocOffset(pub u64);
+
+/// The `DebugLoc` struct represents the DWARF strings
+/// found in the `.debug_loc` section.
+#[derive(Debug, Clone, Copy)]
+pub struct DebugLoc<'input, Endian>
+    where Endian: Endianity
+{
+    debug_loc_section: EndianBuf<'input, Endian>,
+}
+
+impl<'input, Endian> DebugLoc<'input, Endian>
+    where Endian: Endianity
+{
+    /// Construct a new `DebugLoc` instance from the data in the `.debug_loc`
+    /// section.
+    ///
+    /// It is the caller's responsibility to read the `.debug_loc` section and
+    /// present it as a `&[u8]` slice. That means using some ELF loader on
+    /// Linux, a Mach-O loader on OSX, etc.
+    ///
+    /// ```
+    /// use gimli::{DebugLoc, LittleEndian};
+    ///
+    /// # let buf = [0x00, 0x01, 0x02, 0x03];
+    /// # let read_debug_loc_section_somehow = || &buf;
+    /// let debug_loc = DebugLoc::<LittleEndian>::new(read_debug_loc_section_somehow());
+    /// ```
+    pub fn new(debug_loc_section: &'input [u8]) -> DebugLoc<'input, Endian> {
+        DebugLoc { debug_loc_section: EndianBuf(debug_loc_section, PhantomData) }
+    }
+
+    /// Iterate over the `LocationListEntry`s starting at the given offset.
+    ///
+    /// The `address_size` must be match the compilation unit for this location list.
+    /// The `base_address` should be obtained from the `DW_AT_low_pc` attribute in the
+    /// `DW_TAG_compile_unit` entry for the compilation unit that contains this location
+    /// list.
+    ///
+    /// Can be [used with
+    /// `FallibleIterator`](./index.html#using-with-fallibleiterator).
+    pub fn locations(&self,
+                     offset: DebugLocOffset,
+                     address_size: u8,
+                     base_address: u64)
+                     -> Result<LocationListIter<Endian>> {
+        let offset = offset.0 as usize;
+        if self.debug_loc_section.len() < offset {
```

Comment:
Not sure. It won't panic, but the iterator will immediately return None. I'm also not sure if we should make the iterator more strict and require the list end entry, in which case it wouldn't matter that this was <.

We currently use < nearly everywhere for this type of check (and we're missing the check in some places, eg debug_line, debug_abbrev).